### PR TITLE
Add a unit test to cover the unified setting's registration file.

### PR DESCRIPTION
**Pull request by Cosifne**
This is a test-only change.
Originally mentioned by Sam somewhere we need some tests to guard the registration files of Unified Settings.
This PR adds unit test to cover the C# Intellisense page.
It covers
1. Type
2. Default value
3. The storage path to `ISettingManager`. Make sure the value is correct and matches https://github.com/dotnet/roslyn/blob/7bf018744dc7a277070101dcbc4abf2a47b82dd2/src/VisualStudio/Core/Def/Options/VisualStudioOptionStorage.cs#L116
4. The enum to int mapping if the option type is Enum

This would ensure we don't make simple mistakes like typo in the registration file.

The general test code is put under `Microsoft.VisualStudio.LanguageServices.UnitTests`, and the real test is put under `Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests`, so that later we can onboard the VB tests easily.

### CODE REVIEW

**Comment by Cosifne** [comment]
Kindly tag @sharwell  to review.

File: src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            CompletionOptionsStorage.ShowNewSnippetExperienceUserOption
+        );
+
+        internal override ImmutableDictionary<IOption2, object> OptionsToDefaultValue => ImmutableDictionary<IOption2, object>.Empty.
+            Add(CompletionOptionsStorage.SnippetsBehavior, SnippetsRule.AlwaysInclude).
+            Add(CompletionOptionsStorage.EnterKeyBehavior, EnterKeyRule.Never).
+            Add(CompletionOptionsStorage.TriggerOnDeletion, false).
+            Add(CompletionOptionsStorage.ShowItemsFromUnimportedNamespaces, true).
+            Add(CompletionViewOptionsStorage.EnableArgumentCompletionSnippets, false).
+            Add(CompletionOptionsStorage.ShowNewSnippetExperienceUserOption, false);
```
Comment (sharwell) [inline src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs:42]: Why include these? The default values are defined as part of the option, so it doesn't seem necessary to duplicate that.

File: src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        internal override ImmutableDictionary<IOption2, ImmutableArray<object>> EnumOptionsToValues => ImmutableDictionary<IOption2, ImmutableArray<object>>.Empty.
+                Add(CompletionOptionsStorage.SnippetsBehavior, ImmutableArray.Create<object>(SnippetsRule.NeverInclude, SnippetsRule.AlwaysInclude, SnippetsRule.IncludeAfterTypingIdentifierQuestionTab)).
+                Add(CompletionOptionsStorage.EnterKeyBehavior, ImmutableArray.Create<object>(EnterKeyRule.Never, EnterKeyRule.AfterFullyTypedWord, EnterKeyRule.Always));
```
Comment (sharwell) [inline src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs:46]: Can we not just use `Enum.GetValues` for these?

File: src/VisualStudio/Core/Test/UnifiedSettings/UnifiedSettingsTests_UnifiedSettingsStorage.vb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+' See the LICENSE file in the project root for more information.
+
+Imports Microsoft.CodeAnalysis
+
+Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.UnifiedSettings
+    Partial Public Class UnifiedSettingsTests
+
+        ' Mapping from the config name to its path in Unified Settings registration file
+        Private Shared ReadOnly s_unifiedSettingsStorage As New Dictionary(Of String, UnifiedSettingsStorage)() From {
+            {"dotnet_trigger_completion_on_typing_letters", New UnifiedSettingsStorage("textEditor.%LANGUAGE%.intellisense.triggerCompletionOnTypingLetters")},
```
Comment (sharwell) [inline src/VisualStudio/Core/Test/UnifiedSettings/UnifiedSettingsTests_UnifiedSettingsStorage.vb:12]: 💭 Is this mapping not anywhere else?

File: src/VisualStudio/Core/Test/UnifiedSettings/UnifiedSettingsTests.vb [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        ' But in unified settings we always use the correct value for language.
+        ' Use this dictionary to indicate that value in unit test.
+        Friend MustOverride ReadOnly Property OptionsToDefaultValue As ImmutableDictionary(Of IOption2, Object)
+
+        ' The default value of some enum options is overridden at runtime. And it's not shown in the unified settings page.
+        ' Use this dictionary to list all the possible enum values in settings page.
+        Friend MustOverride ReadOnly Property EnumOptionsToValues As ImmutableDictionary(Of IOption2, ImmutableArray(Of Object))
+
+        Protected Sub TestIntelliSensePageSettings(registrationJsonObject As JObject)
+            For Each onboardedOption In OnboardedOptions
```
[the remark this answers is not in the document]
Comment (Cosifne) [reply src/VisualStudio/Core/Test/UnifiedSettings/UnifiedSettingsTests.vb]: Done

File: src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs (hunk 1 above)
Comment (Cosifne) [reply src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs:43]: It's not always true. In many cases we override the default value at runtime, but in unified settings the default value are always put in the registration json.
Case 1:
`CompletionOptionsStorage.TriggerOnDeletion`  Its default value is null, the real default values are set different for [VB](https://github.com/dotnet/roslyn/blob/efdb56597abbe7dff8f5df311f60375877f0b17d/src/VisualStudio/VisualBasic/Impl/Options/IntelliSenseOptionPageControl.xaml.vb#L18) and [C#](https://github.com/dotnet/roslyn/blob/efdb56597abbe7dff8f5df311f60375877f0b17d/src/VisualStudio/CSharp/Impl/Options/IntelliSenseOptionPageControl.xaml.cs#L23). 
The gap here I think is `PerLanguageOption` can only have one default value, but in reality sometimes we needs different default value for each lang.

Case 2: 
`CompletionOptionsStorage.ShowNewSnippetExperienceUserOption` Its default value is null, we use the feature flag trick, the real effect value if `false

File: src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs (hunk 2 above)
Comment (Cosifne) [reply src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs:47]: Use `SnippetsRule` as example here.
https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/Completion/SnippetsRule.cs
It contains four values. And the Default value is `default`.
But it is overridden differently for[ C#](https://github.com/dotnet/roslyn/blob/efdb56597abbe7dff8f5df311f60375877f0b17d/src/Features/CSharp/Portable/Completion/CSharpCompletionService.cs#L55) and [VB ](https://github.com/dotnet/roslyn/blob/efdb56597abbe7dff8f5df311f60375877f0b17d/src/Features/VisualBasic/Portable/Completion/VisualBasicCompletionService.vb#L63) later, 

So, in reality the `default` value is not used, and they are not listed in the unified settings regirstration file.
We just use the real value for each language.
image

So the array here to just to exclude `default` from `Enum.GetValues`

File: src/VisualStudio/Core/Test/UnifiedSettings/UnifiedSettingsTests_UnifiedSettingsStorage.vb (hunk 3 above)
Comment (Cosifne) [reply src/VisualStudio/Core/Test/UnifiedSettings/UnifiedSettingsTests_UnifiedSettingsStorage.vb]: No, not now at least.

Comment (Cosifne) [reply src/VisualStudio/Core/Test/UnifiedSettings/UnifiedSettingsTests_UnifiedSettingsStorage.vb]: They might be moved to https://github.com/dotnet/roslyn/blob/33977308747bf7675931788cec11b78343edc92d/src/VisualStudio/Core/Def/Options/VisualStudioOptionStorage.cs#L116 in future
Unified Settings provides `Microsoft.VisualStudio.Utilities.UnifiedSettings.ISettingsManager`, which would replace the usage of all old `ISettingsManager2` `IVsFeatureFlag` and etc... but it's not required now.

File: src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs (hunk 2 above)
Comment (sharwell) [reply src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs:47]: Yes, I would prefer to use `Enum.GetValues` and then explicitly remove options that are known to be missing. Each exclusion should have its exclusion reason documented.

File: src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs (hunk 1 above)
Comment (Cosifne) [reply src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs]: Done

File: src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs (hunk 2 above)
Comment (Cosifne) [reply src/VisualStudio/CSharp/Test/UnifiedSettings/CSharpUnifiedSettingsTests.cs]: So I use a similar with like https://github.com/dotnet/roslyn/pull/72946#discussion_r1559449317 to address this.

File: src/VisualStudio/Core/Test/Microsoft.VisualStudio.LanguageServices.UnitTests.vbproj [hunk 5]
```diff
@@ -7,6 +7,11 @@
     <UseWpf>true</UseWpf>
     <RootNamespace></RootNamespace>
   </PropertyGroup>
+  <ItemGroup>
+    <Compile Remove="UnifiedSettings\**" />
+    <EmbeddedResource Remove="UnifiedSettings\**" />
+    <Page Remove="UnifiedSettings\**" />
+  </ItemGroup>
```
Comment (sharwell) [inline src/VisualStudio/Core/Test/Microsoft.VisualStudio.LanguageServices.UnitTests.vbproj:14]: Is this necessary? It seems like these all reference a location which does not exist.

**Review by sharwell — APPROVED** [review]

File: src/VisualStudio/Core/Test/Microsoft.VisualStudio.LanguageServices.UnitTests.vbproj (hunk 5 above)
Comment (Cosifne) [reply src/VisualStudio/Core/Test/Microsoft.VisualStudio.LanguageServices.UnitTests.vbproj:14]: It should be removed ; ) VS Auto create this for me

**Pull request merged**